light/provider/http/http.go

```diff
@@ -180,6 +180,15 @@ func (p *http) signedHeader(ctx context.Context, height *int64) (*types.SignedHe
 		commit, err := p.client.Commit(ctx, height)
 		switch {
 		case err == nil:
+			// See https://github.com/cometbft/cometbft/issues/575
+			// If the node is starting at a non-zero height, but does not yet
+			// have any blocks, it can return an empty signed header without
+			// returning an error.
+			if commit.SignedHeader.IsEmpty() {
```

Comment:
I'm not convinced by this. Here, we are changing behavior in production.
The way a light client should work is propagating results from full nodes (after verifying them).
So, to me, the expected behavior from a light client is that it should reply the same as the full nodes that are backing it.
Is it possible to "coalesce" `ErrHeightTooHigh` and _empty signed header_ (my understanding of this change) in testing code rather than production code?
Please let me know if you would like to discuss this synchronously.

Reply:
The question I would ask here is, is the current behaviour in production (panicking when receiving, what appears to be, a valid response from our RPC) correct? I don't think the behaviour's correct. I think we're hitting a corner case here that wasn't thought of previously, i.e. the node starting from a non-zero height, but not yet having started catching up (i.e. a bug).

Several other endpoints also return empty OK responses if the node doesn't yet have data. See the following cases:

https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/rpc/core/blocks.go#L99

https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/rpc/core/blocks.go#L114

https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/rpc/core/blocks.go#L132

https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/rpc/core/blocks.go#L142

Reply:
I took a closer look at this. I was under the impression that the modified method (`signedHeader`) was the implementation of a public RPC endpoint, but it is not the case, it's only used by method `LightBlock`.
So, even if coalescing "empty header" with `ErrHeightTooHigh` feels like we're making the caller of `signedHeader` lose info, in all fairness I don't see a better way of handling the panic we're hitting (I mean, better than what is proposed here).
So this convinces me that this change makes sense. Nevertheless, you might consider moving the `IsEmpty()` check to line [79](https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/light/provider/http/http.go#L79), and have `LightBlock` return a specific error, similarly to what is done in line [82](https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/light/provider/http/http.go#L82).
Thanks for your patience.

Reply:
Not a problem!

The meaning associated with an empty response is: the chain has no data yet. This is the same response that would be returned if the light client requested a header for height 1 but the chain is at height 0. This particular arm of the switch statement is triggered in such a case (i.e. querying for height 1's data when the chain has no data yet): https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/light/provider/http/http.go#L185

The error tells the light client that this is effectively a transient problem and the peer should not be removed as a provider, as per: https://github.com/cometbft/cometbft/blob/5d8789647b6759a3213f078c28279b0d16d1c754/light/provider/errors.go#L9-L11

In the case that I'm catering for here, it's only possible because in our E2E tests we start the chain at height 1000, but it doesn't yet have any blocks. The RPC server handles these two cases differently - if the chain's at height 0 and the request is for height 1, it returns an error; if the chain's at height 1000 and it has no data, it returns an empty OK response.

So our options here are to:
1. Handle the possibility of getting an empty OK response in the light client.
2. Change the behaviour of the RPC to return an error when the chain has no data, regardless of the base height of the chain.
3. Both 1 and 2.

I'm not sure what side effects option (2) would have though.

Reply:
What's the benefit of distinguishing between the two errors? Would the light client knowing which of the two errors occurred make dealing with the error more actionable in some way? (It's only visible to the light client, and light clients generally don't have any control/influence over the nodes to which they're connecting)

To me the impact/effect of both errors is precisely the same so far, but I'm always open to being convinced otherwise.

Reply:
Let's merge this now to at least remove the panic, and then I can follow up with a separate PR if there's value in distinguishing the error into two different errors?

Reply:
Sorry, long time to get back to this. I'm OK with leaving it as is. A common pattern I developed is to try to give as much info in error as possible, to allow (manual) troubleshooting. So, usually, I try to keep different error separate if possible (so a human troubleshooting can distinguish them if needed).
But again, just a suggestion. The PR is good as merged.